src/Common/QueryProfiler.cpp

```
@@ -85,6 +85,92 @@ namespace ErrorCodes
+                                "'timer_create' returned non-zero but didn't set errno. This is bug in your OS.");
+
+            throwFromErrno("Failed to create thread timer", ErrorCodes::CANNOT_CREATE_TIMER);
+        }
+        timer_id.emplace(local_timer_id);
+    }
+}
+
+void Timer::set(UInt32 period)
+{
+    /// Too high frequency can introduce infinite busy loop of signal handlers. We will limit maximum frequency (with 1000 signals per second).
+    if (period < 1000000)
+        period = 1000000;
+    /// Randomize offset as uniform random value from 0 to period - 1.
+    /// It will allow to sample short queries even if timer period is large.
+    /// (For example, with period of 1 second, query with 50 ms duration will be sampled with 1 / 20 probability).
+    /// It also helps to avoid interference (moire).
+    UInt32 period_rand = std::uniform_int_distribution<UInt32>(0, period)(thread_local_rng);
+
+    struct timespec interval{.tv_sec = period / TIMER_PRECISION, .tv_nsec = period % TIMER_PRECISION};
+    struct timespec offset{.tv_sec = period_rand / TIMER_PRECISION, .tv_nsec = period_rand % TIMER_PRECISION};
+
+    struct itimerspec timer_spec = {.it_interval = interval, .it_value = offset};
+    if (timer_settime(*timer_id, 0, &timer_spec, nullptr))
+        throwFromErrno("Failed to set thread timer period", ErrorCodes::CANNOT_SET_TIMER_PERIOD);
+}
+
+void Timer::stop()
+{
+    if (timer_id)
+    {
+        struct timespec stop_timer{.tv_sec = 0, .tv_nsec = 0};
+        struct itimerspec timer_spec = {.it_interval = stop_timer, .it_value = stop_timer};
+        int err = timer_settime(*timer_id, 0, &timer_spec, nullptr);
+        if (err)
+            LOG_ERROR(log, "Failed to stop query profiler timer {}", errnoToString());
+        chassert(!err && "Failed to stop query profiler timer");
+    }
+}
+
+Timer::~Timer()
+{
+    try {
+        cleanup();
+    } catch(...) {}
```

Hiding exceptions is not a good idea, consider using something like this:

```suggestion
    try
    {
        cleanup();
    }
    catch(...)
    {
        tryLogCurrentException(log);
    }
```